### PR TITLE
Fix for #8662 paint_haunted_house_structure crash

### LINKED ISSUE #8662: Crash in paint_haunted_house_structure
Commit:  381e2c1 

Park extracted from https://github.com/OpenRCT2/OpenRCT2/issues/8517#issuecomment-450034859

[park-8517.sv6.gz](https://github.com/OpenRCT2/OpenRCT2/files/2824748/park-8517.sv6.gz)

```
#0  0x0000555555eba745 in paint_haunted_house_structure (session=0x5555563c5be0 <gPaintSession>, rideIndex=11 '\v', direction=1 '\001', xOffset=-32 '\340', yOffset=32 ' ', part=4 '\004', height=179) at ../src/openrct2/ride/gentle/HauntedHouse.cpp:44
#1  0x0000555555eba56e in paint_haunted_house (session=0x5555563c5be0 <gPaintSession>, rideIndex=11 '\v', trackSequence=6 '\006', direction=1 '\001', height=176, tileElement=0x5555564e6cc0 <gTileElements+263952>) at ../src/openrct2/ride/gentle/HauntedHouse.cpp:115
#2  0x0000555555b1e69e in track_paint (session=0x5555563c5be0 <gPaintSession>, direction=1 '\001', height=176, tileElement=0x5555564e6cc0 <gTileElements+263952>) at ../src/openrct2/ride/TrackPaint.cpp:2218
#3  0x0000555555a88ecc in sub_68B3FB (session=0x5555563c5be0 <gPaintSession>, x=2496, y=2400) at ../src/openrct2/paint/tile_element/Paint.TileElement.cpp:301
#4  0x0000555555a8861c in tile_element_paint_setup (session=0x5555563c5be0 <gPaintSession>, x=2496, y=2400) at ../src/openrct2/paint/tile_element/Paint.TileElement.cpp:58
#5  0x0000555555a7d57b in paint_session_generate (session=0x5555563c5be0 <gPaintSession>) at ../src/openrct2/paint/Paint.cpp:218
#6  0x00005555559de8b7 in viewport_paint_column (dpi=0x7fffffffd0d0, viewFlags=1024, sessions=0x0) at ../src/openrct2/interface/Viewport.cpp:929
#7  0x00005555559de811 in viewport_paint (viewport=0x5555563b6030 <g_viewport_list>, dpi=0x7fffffffd220, left=-946, top=2236, right=214, bottom=2276, sessions=0x0) at ../src/openrct2/interface/Viewport.cpp:911
#8  0x00005555559de410 in viewport_render (dpi=0x7fffffffd220, viewport=0x5555563b6030 <g_viewport_list>, left=-946, top=2236, right=214, bottom=2276, sessions=0x0) at ../src/openrct2/interface/Viewport.cpp:826
#9  0x00005555558be314 in window_main_paint (w=0x55555c1cbef0, dpi=0x7fffffffd220) at ../src/openrct2-ui/windows/Main.cpp:89
#10 0x00005555559e765d in window_event_paint_call (w=0x55555c1cbef0, dpi=0x7fffffffd220) at ../src/openrct2/interface/Window.cpp:1594
#11 0x00005555559e6920 in window_draw_single (dpi=0x7fffffffd220, w=0x55555c1cbef0, left=0, top=656, right=1160, bottom=696) at ../src/openrct2/interface/Window.cpp:1276
#12 0x00005555559e63b1 in window_draw (dpi=0x7fffffffd300, w=0x55555c1cbef0, left=0, top=656, right=1160, bottom=696) at ../src/openrct2/interface/Window.cpp:1160
#13 0x00005555559e8f05 in window_draw_all (dpi=0x555556bf6b78, left=0, top=656, right=1160, bottom=696) at ../src/openrct2/interface/Window.cpp:2119
#14 0x00005555559e02f7 in viewport_shift_pixels (dpi=0x555556bf6b78, window=0x55555c1cbef0, viewport=0x5555563b6030 <g_viewport_list>, x_diff=366, y_diff=2167) at ../src/openrct2/interface/Viewport.cpp:408
#15 0x00005555559ddddd in viewport_move (x=-946, y=1580, w=0x55555c1cbef0, viewport=0x5555563b6030 <g_viewport_list>) at ../src/openrct2/interface/Viewport.cpp:499
#16 0x00005555559dd5ff in viewport_update_position (window=0x55555c1cbef0) at ../src/openrct2/interface/Viewport.cpp:631
#17 0x000055555583da56 in WindowManager::SetMainView (this=0x5555569bbf30, x=-366, y=1928, zoom=0, rotation=0) at ../src/openrct2-ui/WindowManager.cpp:499
#18 0x00005555559534b3 in game_load_init () at ../src/openrct2/Game.cpp:1174
#19 0x0000555555940c64 in OpenRCT2::Context::LoadParkFromStream (this=0x555556b564f0, stream=0x7fffffffd828, path="/home/janisozaur/Downloads/park-8517.sv6", loadTitleScreenFirstOnFail=true) at ../src/openrct2/Context.cpp:544
#20 0x0000555555940760 in OpenRCT2::Context::LoadParkFromFile (this=0x555556b564f0, path="/home/janisozaur/Downloads/park-8517.sv6", loadTitleScreenOnFail=true) at ../src/openrct2/Context.cpp:497
#21 0x000055555594645b in OpenRCT2::Context::Launch (this=0x555556b564f0) at ../src/openrct2/Context.cpp:718
#22 0x000055555593f749 in OpenRCT2::Context::RunOpenRCT2 (this=0x555556b564f0, argc=2, argv=0x7fffffffdb58) at ../src/openrct2/Context.cpp:227
#23 0x00005555558283e7 in main (argc=2, argv=0x7fffffffdb58) at ../src/openrct2-ui/Ui.cpp:60
```

https://github.com/OpenRCT2/OpenRCT2/blob/381e2c12bcf4b61390ff6bd97f52891f35bde4a6/src/openrct2/ride/gentle/HauntedHouse.cpp#L44

### PR DESCRIPTION
**Pull request by Xkeeper0**
This series of commits fixes #8662 (as well as two additional crashes that came up after resolving the original one).

There are four commits; the first three abort if `rideEntry` is `NULL`, the final one instead avoids calls that would have used the `baseImageId` derived from it.

I'm not sure exactly what this park is _supposed_ to look like (and it seems like it's still missing things), but at least you can look around and it doesn't crash any more.

2019-02-05-11 11 12-openrct2

### CODE REVIEW

**Comment by duncanspumpkin** [comment]
In general I was under the impression NE uses haunted house code to force rides to not draw. Is there any park that they expect a different drawing when the rideEntry is bad? I'd rather not litter the codebase with big if's like this if its not even used by anyone.

**Comment by duncanspumpkin** [comment]
Having a look at the haunted house code the part you have enabled doesn't actually do anything. All it does is set the interaction mode to vehicle then back to ride. As there is no drawing the interaction mode change doesn't do anything.
 
For the ferris wheel it does make a difference as it draws the A frame of the ride. Unsure if this is actually a useful feature for the user though. Its up for discussion.

**Comment by IntelOrca** [comment]
@duncanspumpkin If `get_ride_entry` can return nullptr then there should always be a check. The only way to reduce that is by passing by reference more, e.g. `rct_ride_entry&`.

File: src/openrct2/world/MapAnimation.cpp [hunk 1]
```diff
@@ -185,6 +185,9 @@ static bool map_animation_invalidate_small_scenery(int32_t x, int32_t y, int32_t
             continue;
 
         sceneryEntry = tileElement->AsSmallScenery()->GetEntry();
+        if (sceneryEntry == nullptr)
+            continue;
```
Comment (duncanspumpkin) [inline src/openrct2/world/MapAnimation.cpp]: Alternatively you could have returned true here as there should not be another small_scenery element at this location. But then NE does like to do funny things like overlapping scenery so maybe this does make more sense.

**Comment by Xkeeper0** [comment]
I was under the impression that this change would not affect anything in use, at least on most multiplayer servers, because setting rides to "Haunted House" _doesn't_ cause a crash -- so `rideEntry` isn't `NULL`. (Setting a ride to that will still make it hidden when it would before, as well...)

> In general I was under the impression NE uses haunted house code to force rides to not draw.

This is an odd case. This map seems to be [The Bayou](https://www.nedesigns.com/park/3746/the-bayou/), but heavily corrupted? In the save file directly from the site, the haunted house actually _is_ a haunted house:
image

_All_ of the rides in #8662's attached save file are corrupted in a way that window for them doesn't show _anything_ other than the view, and a lot of things seem to be either missing or broken compared to the original save file above.

**Comment by Xkeeper0** [comment]
Done.

**Pull request merged**